scipy/sparse/_coo.py

```diff
@@ -1275,6 +1262,460 @@ def _broadcast_to(self, new_shape, copy=False):
 
         return coo_array((new_data, new_coords), new_shape)
 
+    def _process_arrays_for_comparison(self, other, op_name):
+        if is_pydata_spmatrix(other):
+            return NotImplemented
+
+        if not (issparse(other) or isdense(other) or isscalarlike(other)):
+            # If it's a list or whatever, treat it like an array
+            other_a = np.asanyarray(other)
+
+            if other_a.ndim == 0 and other_a.dtype == np.object_:
+                return NotImplemented
+
+            try:
+                other.shape
+            except AttributeError:
+                other = other_a
+
+        if self.ndim < 3 and (isscalarlike(other) or other.ndim < 3):
+            return getattr(_data_matrix, op_name)(self.tocsr(), other)
+
+        # Scalar other.
+        if isscalarlike(other):
+            result_shape = self.shape
+            self = self.reshape(1,-1)
+            result = getattr(_data_matrix, op_name)(self.tocsr(), other)
+            if isinstance(result, sparray):
+                result = result.tocoo()
+            if isinstance(result, np.ndarray | sparray):
+                result = result.reshape(result_shape)
+            return result
+
+        elif isdense(other) or issparse(other):
+            if self.shape != other.shape:
+                raise ValueError(f'Incompatible shapes {self.shape} and {other.shape}')
+            result_shape = self.shape
+
+            # reshaping n-D arrays to 2-D arrays
+            self = self.reshape(1,-1)
+            other = other.reshape(1,-1)
+
+            # routing via 2-D CSR
+            result = getattr(_data_matrix, op_name)(self.tocsr(), other)
+
+            if isinstance(result, sparray):
+                result = result.tocoo()
+
+            # reshaping back to n-D if output is 2-D boolean array
+            if isinstance(result, np.ndarray | sparray):
+                result = result.reshape(result_shape)
+
+            return result
+        else:
+            return NotImplemented
+
+    def __eq__(self, other):
+        return self._process_arrays_for_comparison(other, '__eq__')
```

Comment:
Let's change this to use operator functions instead, as that way we can avoid getattr on the base class.